### PR TITLE
[SPARK-14670][SQL] Allow updating SQLMetrics on driver

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?

On the SparkUI right now we have this SQLTab that displays accumulator values per operator. However, it only displays metrics updated on the executors, not on the driver. It is useful to also include driver metrics, e.g. broadcast time.

After this patch you can update metrics defined in `SparkPlan.metrics` in `doConsume` (if whole stage codegen is enabled) or `doExecute`.
## How was this patch tested?

`SQLListenerSuite`

### CODE REVIEW

**Comment by andrewor14** [comment]
@davies @zsxwing

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 1]
```diff
@@ -52,8 +52,14 @@ private[sql] object SQLExecution {
         try {
           body
         } finally {
+          val driverAccumUpdates = queryExecution.executedPlan
+            .collect { case plan: SparkPlan => plan }
+            .flatMap(_.metrics.values.toSeq)
+            .map { a => a.toInfo(Some(a.localValue), None) }
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: Should we filter out the metrics that have not be updated (zero value)?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: ping

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 2]
```diff
@@ -38,7 +38,13 @@ case class SparkListenerSQLExecutionStart(
...
 
 @DeveloperApi
-case class SparkListenerSQLExecutionEnd(executionId: Long, time: Long)
+case class SparkListenerSQLExecutionEnd(
+    executionId: Long,
+    time: Long,
+    // This allows the driver to update an accumulator's value and have it reported
+    // on the SQL UI. The accumulator needs to be part of SparkPlan.metrics and updated
+    // in doExecute or doConsume, depending on whether whole stage codegen is enabled.
+    driverAccumUpdates: Seq[AccumulableInfo] = Seq.empty[AccumulableInfo])
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: As this class uses the default parameter, SparkListenerSQLExecutionEnd only has one constructor: `this(executionId, time, driverAccumUpdates)`. Recovering from old JSON string will set `driverAccumUpdates` to `null`.

Hence, we cannot use the default parameter here. Instead, we should add two constructors manually like this:
```Scala
@DeveloperApi
case class SparkListenerSQLExecutionEnd(
    executionId: Long,
    time: Long,
    // This allows the driver to update an accumulator's value and have it reported
    // on the SQL UI. The accumulator needs to be part of SparkPlan.metrics and updated
    // in doExecute or doConsume, depending on whether whole stage codegen is enabled.
    driverAccumUpdates: Seq[AccumulableInfo])
  extends SparkListenerEvent {

  def this(executionId: Long, time: Long) {
    this(executionId, time, Seq.empty)
  }
}

object SparkListenerSQLExecutionEnd {
  def apply(executionId: Long, time: Long): SparkListenerSQLExecutionEnd = {
    new SparkListenerSQLExecutionEnd(executionId, time)
  }
}
```

Could you also add a test like this to make sure we don't break the compatibility?
```Scala
    val event = SparkListenerSQLExecutionEnd(1L, 123L)
    val json = JsonProtocol.sparkEventToJson(event)
    val oldJson = json.removeField({ _._1 == "driverAccumUpdates" })
    assert(event === JsonProtocol.sparkEventFromJson(oldJson))
```

**Comment by zsxwing** [comment]
What will happen if reusing a DataFrame? `localValue` in the driver will be a sum of all jobs if a DataFrame?

**Comment by davies** [comment]
@andrewor14 @cloud-fan  Can we target this one to 2.0? The metrics of BroadcastExchange depends on this one.

**Comment by andrewor14** [comment]
OK, I'll bring this up to date

**Comment by cloud-fan** [comment]
I have the same concern with @zsxwing , what if we execute a DataFrame multiple times?

**Comment by andrewor14** [comment]
I won't have time to update this in the this week. I can re-open it next week but if one of you want to take it over before that feel free to do so.

**Pull request closed without merging**